### DISCUSSION TITLE
Un-rendered fields values disappears even when using defaultValues

### DISCUSSION THREAD

**Discussion by samuelpetroline** [react-hook-form/react-hook-form#2781] [Q&A]
First of all, thanks for the library, has been really nice to use.

I don't know exactly how to formulate my problem but here it is:

1. I have a form which is initialized with some defaultValues (synchronously)
2. Based on the initial values, some fields are not rendered
3. Because they are not rendered when submitting these values disappears

Something like this: 

```js
if (!provider) {
  <Controller
  control={control} 
  name="test" 
  render={({ onChange, onBlur, value, name }) => (
    <Checkbox
      onBlur={onBlur}
      onChange={e => onChange(e.target.checked)}
      checked={value}
      name={name}
    />
  )}
/>
}
```

`test`would be `undefined`even though it was initialized with `defaultValues`

Any idea why or how to fix this?

Sorry if it was not clear enough

**Comment by bluebill1049** [comment]
can you share a simple csb?

**Reply by dhmacs** [reply]
I have the same issue. Here the csb I'm working on:
https://codesandbox.io/s/test-react-hook-zzqee?file=/App.js

This simulated getting and refreshing the form data from the db. As you can see I set email value from db, but since the Controller that handles it is commented it disappears from the form data. If you uncomment the email controller component then email would not disappear from form data

**Reply by bluebill1049** [reply]
what's the steps to reproduce? there is a lot stuff in your CSB.

**Comment by dhmacs** [comment]
1 - See that `email` field is missing from form data 
![Screen Shot 2020-09-01 at 10 48 19](https://user-images.githubusercontent.com/4640135/91828779-ddb65e80-ec40-11ea-8aa6-6b2565b9001b.png)

2 - Uncomment line 108-121, refresh and now you see `email` field.
![Screen Shot 2020-09-01 at 10 51 18](https://user-images.githubusercontent.com/4640135/91829020-2cfc8f00-ec41-11ea-977c-dc3aa2ba0058.png)

The form state is reset always in the same way (line 69). 

So basically if you reset the form data with something that is not controlled by react hook form, it gets **removed** from the form data. This happens also for stuff like item ids in arrays, which is a piece of data that is not controlled by registered form input

**Reply by bluebill1049** [reply]
thanks for the detail.

**Reply by bluebill1049** [reply]
That's expected behavior, it's uncontrolled form your form state live inside the input. basically default value is not your form default state, it's for filling up your form's defaultValue. hope this makes more sense.

**Reply by dhmacs** [reply]
Understood 👍  imho I think this makes the library harder to use though, as if you have complex form (like I do) it's really hard to come up with some decent merging strategy between the full state and the partial form state. Like being able to track which items in an array have changed, without having the id of the item (since it's not controlled by the form) it might not be manageable.

I need:
  - be able to handle complex/nested form data
  - be able to reset default form data without affecting dirty fields
  - be able to access default values to calculate diffs to send delta updates to api

Do you have an example that accomplishes these?

**Reply by bluebill1049** [reply]
Sorry no, I don't have an example to accomplish those.

Nothing is impossible, I don't think this lib will do everything for you, for complicate use case like what you suggested above, most of the implementation detail it's on your end. certainly, this lib was not made for such a purpose like what you describe above, rather a generic approach. If this is a requirement at work, I am 100% sure I can meet the requirement.

**Comment by andrwo** [comment]
I like to pipe in to say that this issue is really preventing me from using React-Hook-Forms :(

In Formik:

- Whatever you passed in the defaultValue is preserved, even if it is not any named field
- Whatever you set using setValues() is preserved, even if it is not any named field

You are allowed to keep other non-rendered field data in the data 'structure' that is the form (what you get with getValues() ). This makes it easy to organize very complex forms with multiple IDs key metadata that are part of the data that needs to be submitted together with the form data, but are actually not part of the form itself. 

For example, let say I need to submit a form with : `{   package: { fruits: { keyID: 235, value: *as keyed in field* } } }`. If keyID is a fixed metadata with no rendered (registered) form field, then it gets removed by RHF automatically. Using RHF, I will have to reintegrate the keyID back into the structure, which can get tedious (if not impossible) if I actually have different components handling each part of the form tree of a complex form.

I know one proposed solution is that you deep-merge again the structure with the metadata structure, before submission. But this can only work if the whole form is handled in a single component -- if you have complex nested components with each one handling their own section of the form's metadata, merging this at submit point becomes a nightmare.

I really love this library and love the performance, but this is really a big blocker for me. Is there any way this library can support this kind of usage of keeping named items that do not relate to any rendered fields as it is?

Thanks so much for reading!

**Reply by bluebill1049** [reply]
This sounds like `shouldUnregister: false`, but in V7 we are removing the option. instead, if you want to keep an extra state which is not part of the form, you will have to register them. default values in hook form are for the inputs default value. I know it's convenient to put in the state, it's a tradeoff we are here to make. I am not going to stop you from picking the right tool for the right job. If formik works for you, use it those tools are all here to help.

**Reply by andrwo** [reply]
Hi Bill, I wasn't aware of `shouldUnregister: false`, thanks for the tip. But can I check why `shouldUnregister:false` be removed in V7? It sounds like a very good option to leave there that will enable a whole new class of usage, and it sounds like many people are using it. Is there any possibility I can convince you to leave it in?

The reason why I am moving from Formik is that it is too slow on big complex forms. React-hook-forms would be perfect with a 90% fit already, but certain design decisions seems to be frustrating my efforts to move tho RHF :(  I would love to make everything work if I can.

**Reply by andrwo** [reply]
I have just read through the discussion in https://github.com/react-hook-form/react-hook-form/discussions/3714 ... and just like to ask for clarification because I am a bit new to this.

1) Does that mean that since `shouldUnregister : false` option is being removed in V7 and unregistering should be manual, fields in the object that are passed in to defaultValues, but are not part of any rendered field (i.e. not registered), will still persist in the value unless manually unregistered? That means this behaviour is actually like `shouldUnregister : false` as default.

2) Does setValue(path, value) work to insert the value, if path is not in a registered field (i.e. a metadata field)?

I've tested with 6.14.2 with both passing in data via defaultValues (1) and using setValue (2) and it works beautifully. So I am hoping (and praying!) that this will still work in V7...

**Reply by bluebill1049** [reply]
can you move those questions into v7 RFC, I will respond there. keep everything in one place.
